File: Cuarto/Laboratorio/Talleres/RobocodeLabo/src/robotitoviejo/Robotito.java
package robotitoviejo;

import robocode.JuniorRobot;

public class Robotito extends JuniorRobot {

    private CombatStrategy strategy;

    @Override
    public void run() {
        strategy = new EstrategiaEvasiva();
        strategy.init(this);

        while (true) {

            if (this.strategy.readyToAttack()) {
                this.setStrategy(new EstrategiaOfensiva());
                strategy.init(this);
            }

            strategy.onTick(this);
        }
    }

    public void setStrategy(CombatStrategy newStrat) {
        strategy = newStrat;
    }

    public void onScannedRobot() { strategy.onScannedRobot(this); }

    public void onHitWall() {
        strategy.onHitWall(this);
    }

    public void onHitRobot() {
        strategy.onHitRobot(this);
    }

    public void onHitByBullet() {
        strategy.onHitByBullet(this);
    }
}

File: Cuarto/Laboratorio/Talleres/RobocodeLabo/src/robotitoKotlin2/Robotito.kt
package robotitoKotlin2

import robotitoKotlin.CombatStrategy
import robotitoKotlin.EstrategiaEvasiva
import robotitoKotlin.EstrategiaOfensiva

import robocode.JuniorRobot

class Robotito: JuniorRobot() {

    lateinit var strategy: TwoPhaseCombatStrategy

    override fun run(){
        strategy = EstrategiaEvasiva(this)
        var checked = false

        while(true){
            strategy.onTick()

            if(this.strategy.readyToAttack() && !checked){
                this.strategy = EstrategiaOfensiva(this)
                checked = true
            }

        }

    }

    override fun onScannedRobot(){
        strategy.onScannedRobot()
    }

    override fun onHitWall(){
        strategy.onHitWall()
    }

    override fun onHitRobot() {
        strategy.onHitRobot()
    }

    override fun onHitByBullet() {
        strategy.onHitByBullet()
    }

}